java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightClient.java

```diff
@@ -388,6 +389,57 @@ public ExchangeReaderWriter doExchange(FlightDescriptor descriptor, CallOption..
     }
   }
 
+  /**
+   * Wrapper class to wrap the iterator and handle auth failures and retry.
+   * @param <T> The type of iterator.
+   */
+  public static class WrappedFlightIterator<T> implements Iterator<T> {
+    final Supplier<Iterator<T>> supplier;
+    final Iterator<T> itr;
+    final boolean enableRetry;
+
+    /**
+     * WrappedFlightIterator constructor.
+     * @param supplier The iterator supplier.
+     * @param enableRetry Retry if first call fails.
+     */
+    public WrappedFlightIterator(Supplier<Iterator<T>> supplier, boolean enableRetry) {
+      this.supplier = supplier;
+      this.itr = supplier.get();
+      this.enableRetry = enableRetry;
+    }
+
+    @Override
+    public boolean hasNext() {
+      try {
+        return itr.hasNext();
+      } catch (StatusRuntimeException e) {
+        if (enableRetry && e.getStatus().getCode() == Status.Code.UNAUTHENTICATED) {
+          // TODO: Retry Iterator from the last known position.
```

Comment:
I _believe_ you don't have to worry about this since you will either get the first result or UNAUTHENTICATED. (Or else, you'd have a very odd server that explicitly sends an UNAUTHENTICATED after giving you results.)